src/Types/TypeTelepen.php

```diff
@@ -0,0 +1,186 @@
+        } else {
+            $result = $this->encodeNumeric($code);
+        }
+
+        return $result;
+    }
+
+    protected function encodeAlpha($code) : string
+    {
+
+        if (!preg_match('/[ -~]+/', $code)) { // everything from ASCII32-ASCII127
+            throw new InvalidFormatException("Invalid characters in data");
+        }
+
+        $count = 0;
+
+        /* other implementations use the byte-chr-int type equivalence to work 
+         * with array indices in the conversion/lookup table. It's probably
+         * better to be more explicit with php, hence the use of ord and chr here.
+         */
+
+        // begin with start char
+        $dest = $this->telepen_lookup_table[ord(TELEPEN_START_CHAR)];
+
+        for ($i = 0; $i < strlen($code); $i++) {
+            //$ascii_code = ord(substr($code, $i, 1));
+            $ascii_code = ord($code[$i]);
+            $dest .= ($this->telepen_lookup_table[$ascii_code]);
+            $count += $ascii_code;
+        }
+
+        // Now add check and terminator
+        $check_digit = 127 - ($count % 127);
+        if ($check_digit == 127) {
+            $check_digit = 0;
+        }
+
+        $dest .= $this->telepen_lookup_table[ord($check_digit)];
+        $dest .= $this->telepen_lookup_table[ord(TELEPEN_STOP_CHAR)]; // Stop
+
+        return $dest;
+    }
+
+    private function encodeNumeric(string $code) : string
+    {
+
+        /* If input contains non-numeric or X, exit */
+        if (!preg_match('/^[0-9X]+$/', $code)) {
+            throw new InvalidFormatException("Invalid characters in data");
+        }
+
+        /* If input is an odd length, exit */
+        $t = '';
+        if (strlen($code) % 2 > 0) {
+            throw new InvalidFormatException("There must be an even number of digits");
+        }
+
+        $count = 0;
+        $dest = $this->telepen_lookup_table[ord(TELEPEN_START_CHAR)]; // begin with the start character _
+        
+        for ($i = 0; $i < strlen($code); $i += 2) {
+            $c1 = $code[$i];
+            $c2 = $code[$i+1];
+            /* Input nX is allowed, but Xn is not */
+            if ($c1 == 'X') {
+                throw new InvalidFormatException("Invalid position of X in data");
+            }
+            $glyph = null;
+            if ($c2 == 'X') {
+                $glyph = (ord($c1) - ord('0')) + 17;
+            } else {
+                $glyph = ((10 * (ord($c1) - ord('0'))) + (ord($c2) - ord('0'))) + 27;
+            }
+            $count += $glyph;
+            $dest .= $this->telepen_lookup_table[$glyph];
+        }
+
+        $check_digit = 127 - ($count % 127);
+        if ($check_digit == 127) {
+            $check_digit = 0;
+        }
+
+        $dest .= $this->telepen_lookup_table[$check_digit];
+        $dest .= $this->telepen_lookup_table[ord(TELEPEN_STOP_CHAR)]; // Stop
+
+        return $dest;
+    }
+
+    private function createTelepenConversionTable()
+    {
+        $this->telepen_lookup_table = [
```

Comment:
Can this be formatted better? Can you also provide a comment about how this lookup table works?

Reply:
Small amount of explanation added to table creation, and grouped items into consistent rows of 4 items